docs/src/proposals/big-transactions.md

```diff
@@ -0,0 +1,186 @@
+upper bit of the `u8` will never be set for a valid transaction, we can enable
+it to denote whether a transaction should be decoded with the versioned format
+or not.
+
+#### New Transaction Format
+
+```rust
+pub struct VersionedMessage {
+    /// Version of encoded message.
+    /// The max encoded version is 2^7 - 1 due to the ignored upper disambiguation bit
+    pub version: u8,
+    pub header: MessageHeader,
+    /// Number of read-only account inputs specified thru indexes
+    pub num_readonly_indexed_accounts: u8,
+    #[serde(with = "short_vec")]
+    pub account_keys: Vec<Pubkey>,
+    /// All the account indexes used by this transaction
+    #[serde(with = "short_vec")]
+    pub account_indexes: Vec<AccountIndex>,
+    pub recent_blockhash: Hash,
+    /// Compiled instructions stay the same, account indexes continue to be stored
+    /// as a u8 which means the max number of account_indexes + account_keys is 256.
+    #[serde(with = "short_vec")]
+    pub instructions: Vec<CompiledInstruction>,
+}
+
+pub struct AccountIndex {
+    pub account_key_offset: u8,
+    // 1-3 bytes used to lookup address in index account
+    pub index_account_offset: CompactU16,
+}
+```
+
+#### Size changes
+
+- Extra byte for version field
+- Extra byte for number of total account index inputs
+- Extra byte for number of readonly account index inputs
+- Most indexes will be compact and use 2 bytes + index address
+- Cost of each additional index account is ~2 bytes
+
+### Limitations
+
+- Max of 256 accounts may be specified in a transaction because u8 is used by compiled
+instructions to index into transaction message account keys.
+Indexes can hold up to 2^16 keys. Smaller indexes is ok. Each index is then u16
+- Transaction signers may not be specified using an on-chain account index, the
+full address of each signer must be serialized in the transaction. This ensures
+that the performance of transaction signature checks is not affected.
+
+## Security Concerns
+
+### Resource consumption
+
+Enabling more account inputs in a transaction allows for more program
+invocations, write-locks, and data reads / writes. Before indexes are live, we
+need transaction-wide compute limits and increased costs for write locks and
+data reads.
+
+### Front running
+
+If the addresses listed within an index account are modifiable, front running
+attacks could modify which index accounts are accessed from a later transaction.
+For this reason, we propose that any stored address is immutable and that index
+accounts themselves may not be removed.
+
+### Denial of service
+
+Index accounts will be read very frequently and will therefore be a more high
+profile target for denial of service attacks through write locks similar to
+sysvar accounts.
+
+Since stored accounts inside index accounts are immutable, reads and writes
+to index accounts could be parallelized as long as all referenced addresses
+are for indexes less than the current number of addresses stored.
+
```

Comment:
Maybe account index must be grouped by program id and authorized to the program id owner?

I came up with this attack (unrealistic though): if tx uses untrusted index account...:

1. malicious index account owner appends good account index entry at N
2. before 1. is rooted, alice is tricked to create and sign a tx referencing the newly added entry according to commitment=processed
3. malicious colluding leader creates new fork reverting 1.
4. malicous index account owner appends bad account index entry at N (effective replacement)
5. so, alice  is tricked into buying garbage or trading against bad clob. etc?

Reply:
also, it will be a good practice for any daaps to check the actually referenced account address via index before signing like they're doing for account keys?

Otherwise, their signer bit can be abused covertly by referencing via cpi and the compromised account index account.

Reply:
also thought about hard bad possible interplay regarding nonces and gossiped vote transactions but I couldn't come up with anything particularly bad.

Reply:
> Otherwise, their signer bit can be abused covertly by referencing via cpi and the compromised account index account.

Can you please elaborate a bit more here, I didn't quite understand this.

Reply:
> also, it will be a good practice for any wallets to check the actually referenced account address via index before signing like they're doing for account keys?

Hmm, we might want to add an integrity check. Perhaps we add the sha256 hash of all accounts referenced via indexes to the transaction message?

It might be overkill though. This is not really any different from a malicious actor convincing a user to use an account with some state A1 and then forking the chain to change the account to state A2. It's the client's responsibility to confirm the finality all account state before signing.

Reply:
Actually, I think an integrity check could be done as a separate instruction. Users who desire more security can add an assertion instruction which hashes the indexed account addresses together.

Reply:
The effects of this indirection on what the transaction signatures are saying has been bothering me a bit as well. Ideally, we'd sign over the actual addresses, use the indices for broadcast, then recover the addresses before sigverify.  I don't think this is practical though.

Alternatively, we could add another field of `{indexed_addresses_witness_index: u8, indexed_addresses_signature: Signature}`, that's checked at runtime between accounts load and transaction execution.  This should be robust if we require that the `address_witness` has also signed the broadcast encoding. It'll cost us 65 bytes though.

> Hmm, we might want to add an integrity check. Perhaps we add the sha256 hash of all accounts referenced via indexes to the transaction message?

Theoretically, this would be vulnerable to a birthday attack unless the program had full control of the hash.  We can mitigate by mixing in date outside an attackers control, like the most recent blockhash. It would be pretty costly to the size of the index entries though

Attaching an `insertion_slot` to each index and requiring a cool down period  before use would likely be sufficient

> It might be overkill though.

I'm apt to agree.  It seems like these attacks need sufficient control over the network as to assume it's compromised in many other ways.

Reply:
> > also, it will be a good practice for any wallets to check the actually referenced account address via index before signing like they're doing for account keys?
> > 
> > Otherwise, their signer bit can be abused covertly by referencing via cpi and the compromised account index account.
> 
> Can you please elaborate a bit more here, I didn't quite understand this.

I think my concern is no longer valid with these bits :) :

> If the same account is referenced in a transaction by address as well as through
> an index, the transaction should be rejected to avoid conflicts when determining
> if the account is a signer or writeable.

> Transaction signers may not be specified using an on-chain account index, the
> full address of each signer must be serialized in the transaction. This ensures
> that the performance of transaction signature checks is not affected.

Reply:
I added a paragraph about the attack vector where indexes are modified pre-finalization. I think we can just recommend that clients wait for finalization before using an index. They can also add their own integrity checks if needed